File: src/main/java/io/github/m/Bool.java
package io.github.m;

/**
 * M wrapper class for booleans.
 */
public final class Bool implements Value {
    /**
     * The primitive value of this boolean.
     */
    public final boolean value;

    private Bool(boolean value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return Boolean.toString(value);
    }

    /**
     * Partially applies this value to an argument.
     */
    @Override
    public Value invoke(Value arg) {
        return new Partial(this, arg);
    }

    /**
     * Returns the first argument if this is true, or the second if this is false.
     */
    @Override
    public Value invoke(Value arg1, Value arg2) {
        return value ? arg1 : arg2;
    }

    /**
     * Wraps a primitive boolean in a boolean.
     */
    public static Bool valueOf(boolean value) {
        if (value) {
            return TRUE;
        } else {
            return FALSE;
        }
    }

    /**
     * Converts a value to a boolean.
     */
    public static Bool from(Value value) {
        if (value instanceof Bool) {
            return (Bool) value;
        } else {
            return (Bool) value.invoke(TRUE, FALSE);
        }
    }

    /**
     * Converts a value to a primitive boolean.
     */
    public static boolean primitiveFrom(Value value) {
        return Bool.from(value).value;
    }

    /**
     * The singleton truthy value.
     */
    public static final Bool TRUE = new Bool(true);

    /**
     * The singleton falsy value.
     */
    public static final Bool FALSE = new Bool(false);

    @MField(name = "true")
    public static final Value $true = TRUE;

    @MField(name = "false")
    public static final Value $false = FALSE;
}

File: src/main/kotlin/io/github/m/Bool.kt
package io.github.m

/**
 * M wrapper class for bools.
 */
sealed class Bool : Value {
    object True : Bool() {
        override fun invoke(arg: Value) = Value { arg2 -> this(arg, arg2) }
        override fun invoke(arg1: Value, arg2: Value) = arg1
        override fun toString() = "true"
    }

    object False : Bool() {
        override fun invoke(arg: Value) = Value { arg2 -> this(arg, arg2) }
        override fun invoke(arg1: Value, arg2: Value) = arg2
        override fun toString() = "false"
    }

    companion object {
        fun from(value: Value) = value as? Bool ?: value(True, False) as Bool
        operator fun invoke(boolean: Boolean) = if (boolean) True else False
    }

    /**
     * M bool definitions.
     */
    @Suppress("unused")
    object Definitions {
        @MField("true")
        @JvmField
        val `true`: Value = Bool.True

        @MField("false")
        @JvmField
        val `false`: Value = Bool.False
    }
}